File: skills/vector-index/scripts/grounding.py
#!/usr/bin/env python3
"""
grounding — provenance, confidence tiers, and lexical claim verification.

The trust layer generalized from the legal-RAG document (C5/C6): retrieval tools
still let downstream models hallucinate, and self-rated confidence is unreliable
(AUROC ~0.58). So instead we derive confidence from *retrieval strength and signal
agreement*, expose it as a categorical tier (lawyers and engineers both reason
about tiers, not opaque numbers), and offer a verifier that checks whether a
generated claim is actually supported by the spans it cites — flagging the rest.

Stdlib only; the default verifier is lexical, so it needs no model. An embedding
verifier can be layered on later by passing an encoder.
"""

from __future__ import annotations

from hybrid import tokenize

# Tuned for the default cross-encoder (ms-marco-MiniLM logits, ~ -11..+11) when
# reranked, and cosine similarity (0..1) when not. Constants, not magic numbers.
_RERANK_HIGH, _RERANK_MED = 4.0, 0.0
_COSINE_HIGH, _COSINE_MED = 0.55, 0.40


def confidence_tier(results: list[dict], reranked: bool = True) -> str:
    """A High/Medium/Low tier from the top result's score plus signal agreement.

    "Agreement" = a top hit found by BOTH the dense and lexical legs is stronger
    evidence than one leg alone, so it promotes the tier.
    """
    if not results:
        return "low"
    top = results[0].get("rerank_score")
    if top is None:
        top = results[0].get("vector_score")
    top = top if top is not None else 0.0
    agree = any(len(r.get("signals", [])) >= 2 for r in results[:3])
    hi, med = (_RERANK_HIGH, _RERANK_MED) if reranked else (_COSINE_HIGH, _COSINE_MED)
    if top >= hi or (top >= med and agree):
        return "high"
    if top >= med or agree:
        return "medium"
    return "low"


def verify_claim(claim: str, sources: list[str], threshold: float = 0.30) -> dict:
    """Is `claim` grounded in any of `sources`? Lexical recall of claim terms.

    score = |claim_terms ∩ source_terms| / |claim_terms| for the best source.
    Returns {supported, score, source_index}. Cheap, model-free, and language-
    agnostic — a first-pass groundedness gate before any expensive check.
    """
    cterms = set(tokenize(claim))
    if not cterms:
        return {"supported": False, "score": 0.0, "source_index": -1}
    best_score, best_idx = 0.0, -1
    for i, src in enumerate(sources):
        overlap = len(cterms & set(tokenize(src))) / len(cterms)
        if overlap > best_score:
            best_score, best_idx = overlap, i
    return {
        "supported": best_score >= threshold,
        "score": round(best_score, 3),
        "source_index": best_idx if best_score >= threshold else -1,
    }


def signal_summary(results: list[dict]) -> dict:
    """Count how the top results were found — provenance at the result-set level."""
    dense = sum(1 for r in results if "dense" in r.get("signals", []))
    lexical = sum(1 for r in results if "lexical" in r.get("signals", []))
    both = sum(1 for r in results if len(r.get("signals", [])) >= 2)
    return {"dense": dense, "lexical": lexical, "both": both, "total": len(results)}

File: src/search/grounding.ts
/**
 * grounding — provenance, confidence tiers, and lexical claim verification.
 *
 * Confidence is derived from retrieval strength + signal agreement (not from a
 * model's unreliable self-rating) and exposed as a categorical tier. The verifier
 * checks whether a claim is lexically supported by its cited spans. Model-free.
 * Ported from grounding.py.
 */
import type { ConfidenceTier, SearchHit } from '../db/types.ts'

// Tuned for the default cross-encoder (ms-marco logits ~ -11..+11) when reranked,
// and cosine similarity (0..1) when not.
const RERANK_HIGH = 4.0
const RERANK_MED  = 0.0
const COSINE_HIGH = 0.55
const COSINE_MED  = 0.4

const TOKEN_RE = /[0-9]+(?:[/:.\-][0-9]+)+|\w+/gu

export function tokenize (text: string): string[] {
  return text.toLowerCase().match(TOKEN_RE) ?? []
}

/** Loose result shape matching the Python dicts (rerank_score/vector_score/signals). */
export interface GroundingResult {
  rerank_score?: number | null;
  vector_score?: number | null;
  signals?:      string[];
}

/** Faithful port of confidence_tier over loose dict results. */
export function confidenceTierFromResults (
  results: GroundingResult[],
  reranked = true,
): ConfidenceTier {
  if (results.length === 0)
    return 'low'

  let top = results[0].rerank_score ?? results[0].vector_score ?? 0
  if (top == null)
    top = 0

  const agree       = results.slice(0, 3).some(r => (r.signals?.length ?? 0) >= 2)
  const [ hi, med ] = reranked ? [ RERANK_HIGH, RERANK_MED ] : [ COSINE_HIGH, COSINE_MED ]
  if (top >= hi || top >= med && agree)
    return 'high'
  if (top >= med || agree)
    return 'medium'
  return 'low'
}

/** Typed contract over SearchHit[]: uses rerank score when present, else dense. */
export function confidenceTier (hits: SearchHit[]): ConfidenceTier {
  if (hits.length === 0)
    return 'low'

  const reranked                   = hits[0].rerank != null
  const results: GroundingResult[] = hits.map(h => ({
    rerank_score: h.rerank ?? null,
    vector_score: h.dense,
    signals:      [ h.dense > 0 ? 'dense' : '', h.sparse > 0 ? 'lexical' : '' ].filter(Boolean),
  }))
  return confidenceTierFromResults(results, reranked)
}

export interface VerifyClaimResult {
  supported:    boolean;
  score:        number;
  source_index: number;
}

/** Lexical recall of claim terms against the best-matching source. Model-free. */
export function verifyClaim (claim: string, sources: string[], threshold = 0.3): VerifyClaimResult {
  const cterms = new Set(tokenize(claim))
  if (cterms.size === 0)
    return { supported: false, score: 0, source_index: -1 }

  let bestScore = 0
  let bestIdx   = -1
  sources.forEach((src, i) => {
    const sterms = new Set(tokenize(src))
    let inter = 0
    for (const t of cterms)
      if (sterms.has(t))
        inter++

    const overlap = inter / cterms.size
    if (overlap > bestScore) {
      bestScore = overlap
      bestIdx = i
    }
  })
  return {
    supported:    bestScore >= threshold,
    score:        Math.round(bestScore * 1000) / 1000,
    source_index: bestScore >= threshold ? bestIdx : -1,
  }
}

export interface SignalSummary {
  dense:   number;
  lexical: number;
  both:    number;
  total:   number;
}

export function signalSummary (hits: SearchHit[]): SignalSummary {
  const dense   = hits.filter(h => h.dense > 0).length
  const lexical = hits.filter(h => h.sparse > 0).length
  const both    = hits.filter(h => h.dense > 0 && h.sparse > 0).length
  return { dense, lexical, both, total: hits.length }
}
